File: number_theory/chinese_remainder_theorem.cpp
// chinese_remainder_theorem.cpp - Solve system of congruences using CRT.
// Author: Scott

#include <iostream>
#include <vector>
using namespace std;

int mod_inverse(int a, int m) {
    int m0 = m, t, q;
    int x0 = 0, x1 = 1;

    if (m == 1) return 0;

    while (a > 1) {
        q = a / m;
        t = m;

        m = a % m, a = t;
        t = x0;
        x0 = x1 - q * x0;
        x1 = t;
    }

    if (x1 < 0) x1 += m0;
    return x1;
}

int chinese_remainder(vector<int> num, vector<int> rem) {
    int prod = 1;
    for (int n : num) prod *= n;

    int result = 0;
    for (int i = 0; i < num.size(); ++i) {
        int pp = prod / num[i];
        result += rem[i] * mod_inverse(pp, num[i]) * pp;
    }
    return result % prod;
}

int main() {
    vector<int> num = {3, 5, 7};
    vector<int> rem = {2, 3, 2};

    cout << "Solving system: x ≡ 2 mod 3, x ≡ 3 mod 5, x ≡ 2 mod 7" << endl;
    int result = chinese_remainder(num, rem);
    cout << "Solution x = " << result << endl;
    return 0;
}

File: number_theory/chinese_remainder_theorem.py
"""
chinese_remainder_theorem.py - Solve a system of modular congruences using the Chinese Remainder Theorem (CRT).
Author: Scott
"""

def extended_euclidean(a, b):
    if b == 0:
        return a, 1, 0
    g, x1, y1 = extended_euclidean(b, a % b)
    return g, y1, x1 - (a // b) * y1

def chinese_remainder(n, a):
    prod = 1
    for ni in n:
        prod *= ni
    result = 0
    for ni, ai in zip(n, a):
        p = prod // ni
        _, inv, _ = extended_euclidean(p, ni)
        result += ai * inv * p
    return result % prod

if __name__ == "__main__":
    print("Solving system:")
    print("x ≡ a1 mod n1")
    print("x ≡ a2 mod n2")
    print("...")
    n = list(map(int, input("Enter moduli (n) separated by spaces: ") or "3 5 7".split()))
    a = list(map(int, input("Enter remainders (a) separated by spaces: ") or "2 3 2".split()))
    print(f"Solution x ≡ {chinese_remainder(n, a)} mod {eval('*'.join(map(str,n)))}")
